### PR TITLE
Added support for confidential filter in issues

### PR DESCRIPTION
**Pull request by martin-bougie-ubi**
Added support for confidential filter when querying GitLab issues

### CODE REVIEW

File: NGitLab.Mock/Clients/IssueClient.cs [hunk 1]
```diff
@@ -231,6 +231,11 @@ private IEnumerable<Issue> FilterByQuery(IEnumerable<Issue> issues, IssueQuery q
                 issues = issues.Where(i => string.Equals(i.Milestone?.Title, query.Milestone, StringComparison.Ordinal));
             }
 
+            if (query.Confidential != null)
+            {
+                issues = issues.Where(i => query.Confidential.Value || !i.Confidential);
```
Comment (meziantou) [inline NGitLab.Mock/Clients/IssueClient.cs]: The filter should use the value of `query.Confidential`.

- query.Confidential == true => include all issues
- query.Confidential == false => include only non-confidential issues

In any case, if an issue is confidential, it should only be visible if the current user is `>= Reported` on the project or if the user is the author of the issue. (https://docs.gitlab.com/ee/user/project/issues/confidential_issues.html#permissions-and-access-to-confidential-issues)

**Review by meziantou — CHANGES_REQUESTED** [review]

File: NGitLab.Mock/Clients/IssueClient.cs (hunk 1 above)
Comment (martin-bougie-ubi) [reply NGitLab.Mock/Clients/IssueClient.cs]: I added support for filtering on the user in the various gets of the mock. As for the filter, we made an incorrect assumption on the API behaviour. If the filter is set to true, then the API only returns confidential issues. If set to false, it only returns public issues. If the filter is not set, it returns both. We modified the mocks to reflect this behaviour.

File: NGitLab.Mock.Tests/NGitLab.Mock.Tests.csproj [hunk 2]
```diff
@@ -1,7 +1,10 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="Current">
   <PropertyGroup>
     <TargetFrameworks>net5.0;net472</TargetFrameworks>
     <IsPackable>false</IsPackable>
+    <FileUpgradeFlags>40</FileUpgradeFlags>
```
Comment (meziantou) [inline NGitLab.Mock.Tests/NGitLab.Mock.Tests.csproj]: Revert

File: NGitLab.Mock/NGitLab.Mock.csproj [hunk 3]
```diff
@@ -1,10 +1,13 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project>
+<Project ToolsVersion="Current">
   <Import Project="$([System.IO.Path]::Combine($(SolutionDir), '.SwitchToSource', $(SolutionFileName.Replace('.SwitchToSource.sln', '.sln')), 'import', 'NGitLab.Mock.before.props'))" Condition="$(SolutionFileName.EndsWith('.SwitchToSource.sln', StringComparison.OrdinalIgnoreCase)) And Exists($([System.IO.Path]::Combine($(SolutionDir), '.SwitchToSource', $(SolutionFileName.Replace('.SwitchToSource.sln', '.sln')), 'import', 'NGitLab.Mock.before.props')))" />
   <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
   <PropertyGroup>
     <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
     <NoWarn>$(NoWarn);NU5104</NoWarn>
+    <FileUpgradeFlags>40</FileUpgradeFlags>
```
Comment (meziantou) [inline NGitLab.Mock/NGitLab.Mock.csproj]: revert

File: NGitLab/NGitLab.csproj [hunk 4]
```diff
@@ -1,6 +1,9 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="Current">
   <PropertyGroup>
     <TargetFrameworks>net45;netstandard2.0</TargetFrameworks>
+    <FileUpgradeFlags>40</FileUpgradeFlags>
```
Comment (meziantou) [inline NGitLab/NGitLab.csproj]: revert

File: NGitLab.Mock/Clients/IssueClient.cs [hunk 5]
```diff
@@ -106,7 +106,15 @@ public IEnumerable<Models.Issue> ForProject(int projectId)
         {
             using (Context.BeginOperationScope())
             {
-                return GetProject(projectId, ProjectPermission.View).Issues.Select(i => i.ToClientIssue()).ToList();
+                var project = GetProject(projectId, ProjectPermission.View);
+
+                return project
+                    .Issues
+                    .Where(i => project.CanUserViewConfidentialIssues(Context.User)
+                        || i.Author.Id == Context.User.Id
```
Comment (meziantou) [inline NGitLab.Mock/Clients/IssueClient.cs]: I think you can create a new method `Issue.CanUserViewIssue(User)` and use it in all methods

File: NGitLab.Mock/Clients/IssueClient.cs [hunk 6]
```diff
@@ -106,7 +106,15 @@ public IEnumerable<Models.Issue> ForProject(int projectId)
...
+                        || i.Assignee.Id == Context.User.Id)
```
Comment (meziantou) [inline NGitLab.Mock/Clients/IssueClient.cs]: I don't think this is true, https://docs.gitlab.com/ee/user/project/issues/confidential_issues.html#permissions-and-access-to-confidential-issues

> However, a guest user can also create confidential issues, but can only view the ones that they created themselves.

**Review by meziantou — CHANGES_REQUESTED** [review]

File: NGitLab.Mock/Issue.cs [hunk 7]
```diff
@@ -95,5 +95,10 @@ public Models.Issue ToClientIssue()
                 Confidential = Confidential,
             };
         }
+
+        public bool CanUserViewIssue(User user)
+        {
+            return !Confidential || Author.Id == user.Id;
```
Comment (meziantou) [inline NGitLab.Mock/Issue.cs]: Should include `Project.CanUserViewConfidentialIssues` so you can actually simplify all calls

File: NGitLab.Mock/Clients/IssueClient.cs (hunk 6 above)
Comment (martin-bougie-ubi) [reply NGitLab.Mock/Clients/IssueClient.cs]: I removed the free pass for assignees.

File: NGitLab.Mock/Issue.cs (hunk 7 above)
Comment (martin-bougie-ubi) [reply NGitLab.Mock/Issue.cs]: Good point, hadn't realized that the project object was already referenced inside the issue.

**Review by meziantou — APPROVED** [review]

**Pull request merged**